include/kernel.h

```diff
@@ -1387,7 +1387,7 @@ struct k_timer {
 	_OBJECT_TRACING_NEXT_PTR(k_timer)
 };
 
-#define _K_TIMER_INITIALIZER(obj, expiry, stop) \
+#define Z_TIMER_INITIALIZER(obj, expiry, stop) \
```

Comment:
This seems very unintuitive to me. This API operates on an object called k_timer, so whatever the prefix is (or isn't) the name should still contain K_TIMER

Reply:
I disagree.
These are private APIs, this is why the name had a leading underscore. End users shouldn't be using this. The naming convention for private objects is to prefix with z_ instead of k_

Reply:
What about Z_K_TIMER ? Though it's a bit ugly, it does keep the full object type name. Knowing also that 'K' has a strong meaning, unlike Z... unless we rename all the kernel types k_<> to z_<> but that would be even weirder.

Reply:
I'm not opposed to the z_, just that the name of the API should contain the object that it operates on (it's a k_timer not a z_timer).

Btw, what exactly do you mean by "internal"? That only the k_timer code should be using this? If that's the case, how do you propose getting rid of users like this: https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/bluetooth/host/hci_core.c#L72

Reply:
> I'm not opposed to the z_, just that the name of the API should contain the object that it operates on (it's a k_timer not a z_timer).

I don't really have a strong feeling about this one way or another, aesthetically z_k_ seems awkward but it's just a few macros. I'll go with the prevailing opinion.

> Btw, what exactly do you mean by "internal"? That only the k_timer code should be using this? 

I am not sure where our policy is documented on subsys/ with regards to private kernel APIs.

For those macros in particular, IIRC this was due to objects declared with them not being accessible to user mode.

The solution I'm currently pursuing is to implement futexes, they don't any restrictions where their memory can live.

Reply:
Btw, either of these changes will likely increase the chance of application code accessing the internal APIs, since for an uneducated/naive developer who's looking at include/kernel.h a z_* API from a project called Zephyr will not look in anyway suspicious or communicate a "stay away!" message. I realize the ship may already have sailed, but I'd have at least considered putting any internal APIs into a separate `kernel_internal.h`, and if we want to be completely clear to the user throw an `_internal` suffix at them.

Reply:
The leading underscore is the one targeted, any other prefix not starting with it is fine once we agree on it.

Reply:
keep prefix `k` for private kernel API is ok for me. @pfl what do you think ? I can't disagree with @jhedberg points.

Reply:
Fine with me, but you'd need to spell out if you want a subset of the former '_k_*' functions or all of them changed to prefix 'k_'. I'm going to assume all, so I'll update the patch set accordingly.

The difficult part of this discussion is that it is nowhere explicitly stated which functions are actually exported to the rest of the system ("kernel" and "user space"), and which are "kernel" internal. Johan's comment about include/ and kernel/include is spot on, even I wouldn't know the difference.

We probably need to document this and start moving include files into the same hierarchy we now have for ext/, subsys/ etc.. Right now include/ is a bit too messy when just trying to comprehend available features by reading the individual .h files and trying to figure out if they are ext/, subsys/ or something elsel. Going in the direction from Kconfig variables to finding the relevant include files is the way it seems to work now.

Reply:
> keep prefix k for private kernel API is ok for me

No, please do not do this.
These private APIs, which can disappear or change at *any* time, need to be clearly distinguishable from Zephyr's public interfaces!
z_ makes this very clear that this is a different namespace.

Reply:
> These private APIs, which can disappear or change at any time, need to be clearly distinguishable from Zephyr's public interfaces!

@andrewboie I completely agree with this.

> z_ makes this very clear that this is a different namespace.

Indeed, but not that it's a private API. What makes things worse is that currently these are in the same public header file as the public APIs and (at least some) private APIs don't even have a code comment saying they're private. For a naïve reader the private APIs will not look private at all.

Reply:
> What makes things worse is that currently these are in the same public header file as the public APIs and (at least some) private APIs don't even have a code comment saying they're private. For a naive reader the private APIs will not look private at all.

But that is then a bit of a different issue. I'm not happy at all that the private/public/subsys/ext header files are mixed into the same directory, at minimum the public and private APIs should be in different files and the files properly ordered into a hierarchy. That seems to be a task for 1.15, though.

Reply:
@pfl understood, and agreed mostly. I just think that (ignoring the standard compliance, reserved namespace issue) the _ prefix had at least a slightly better chance of screaming "private!" than z_ has.

Reply:
> Indeed, but not that it's a private API. What makes things worse is that currently these are in the same public header file as the public APIs

+1 for splitting public into private APIs into separate headers. That's the best way to keep namespace clean. Except that in our case it will resolve the situation only "so much", as e.g. we have private macros used to define public APIs. Still, it's the best mechanism we'd have to structure and organize Zephyr namespace.

Reply:
> @pfl understood, and agreed mostly. I just think that (ignoring the standard compliance, reserved namespace issue) the _ prefix had at least a slightly better chance of screaming "private!" than z_ has.

I agree here. But as we want to be MISRA C complieant, a prefix of '_' is out of bounds anyway.